pkg/netceptor/external_backend.go

```diff
@@ -16,6 +17,128 @@ type ExternalBackend struct {
 	sessChan chan BackendSession
 }
 
+// MessageConn is an abstract connection that sends and receives whole messages (datagrams)
+type MessageConn interface {
+	WriteMessage(ctx context.Context, data []byte) error
+	ReadMessage(ctx context.Context, timeout time.Duration) ([]byte, error)
+	SetReadDeadline(t time.Time) error
+	Close() error
+}
+
+// netMessageConn implements MessageConn for Go net.Conn
+type netMessageConn struct {
+	conn   net.Conn
+	framer framer.Framer
+}
+
+// MessageConnFromNetConn returns a MessageConnection that wraps a net.Conn
+func MessageConnFromNetConn(conn net.Conn) MessageConn {
+	return &netMessageConn{
+		conn:   conn,
+		framer: framer.New(),
+	}
+}
+
+// WriteMessage writes a message to the connection
+func (mc *netMessageConn) WriteMessage(ctx context.Context, data []byte) error {
+	if ctx.Err() != nil {
+		return fmt.Errorf("session closed: %s", ctx.Err())
+	}
+	buf := mc.framer.SendData(data)
+	n, err := mc.conn.Write(buf)
+	if err != nil {
+		return err
+	}
+	if n != len(buf) {
+		return fmt.Errorf("partial data sent")
+	}
+	return nil
+}
+
+// ReadMessage reads a message from the connection
+func (mc *netMessageConn) ReadMessage(ctx context.Context, timeout time.Duration) ([]byte, error) {
+	buf := make([]byte, MTU)
+	err := mc.conn.SetReadDeadline(time.Now().Add(timeout))
+	if err != nil {
+		return nil, err
+	}
+	for {
+		if ctx.Err() != nil {
+			return nil, fmt.Errorf("session closed: %s", ctx.Err())
+		}
+		if mc.framer.MessageReady() {
+			break
+		}
+		n, err := mc.conn.Read(buf)
+		if n > 0 {
+			mc.framer.RecvData(buf[:n])
+		}
+		if nerr, ok := err.(net.Error); ok && nerr.Timeout() {
+			return nil, ErrTimeout
+		}
+		if err != nil {
+			return nil, err
+		}
+	}
+	buf, err = mc.framer.GetMessage()
+	if err != nil {
+		return nil, err
+	}
+	return buf, nil
+}
+
+// SetReadDeadline sets the deadline by which a message must be read from the connection
+func (mc *netMessageConn) SetReadDeadline(t time.Time) error {
+	panic("implement me")
+}
+
+// Close closes the connection
+func (mc *netMessageConn) Close() error {
+	return mc.conn.Close()
+}
+
+// websocketMessageConn implements MessageConn for Gorilla websocket.Conn
+type websocketMessageConn struct {
+	conn *websocket.Conn
+}
+
+// MessageConnFromWebsocketConn returns a MessageConnection that wraps a Gorilla websocket.Conn
+func MessageConnFromWebsocketConn(conn *websocket.Conn) MessageConn {
+	return &websocketMessageConn{
+		conn: conn,
+	}
+}
+
+// WriteMessage writes a message to the connection
+func (mc *websocketMessageConn) WriteMessage(ctx context.Context, data []byte) error {
+	if ctx.Err() != nil {
+		return fmt.Errorf("session closed: %s", ctx.Err())
+	}
+	return mc.conn.WriteMessage(websocket.BinaryMessage, data)
```

Comment:
Is there a use-case where you might want to be sending text messages?

Reply:
No, because this is a backend connection, so the only thing Receptor will ever send over it is Receptor wire protocol packets, which are binary data structures.